src/vm/types.rs

```diff
@@ -1,66 +1,208 @@
+use std::fmt;
 use std::collections::BTreeMap;
 
-use vm::InterpreterResult;
-use vm::errors::Error;
+use address::c32;
 use vm::representations::SymbolicExpression;
-use vm::{eval, Context, Environment};
+use vm::errors::{Error, InterpreterResult as Result};
+use util::hash;
+
+const MAX_VALUE_SIZE: i128 = 1024 * 1024; // 1MB
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct TupleTypeSignature {
+    type_map: BTreeMap<String, TypeSignature>
+}
 
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub enum AtomTypeIdentifier {
     VoidType,
     IntType,
     BoolType,
-    BufferType,
+    BufferType(u32),
+    PrincipalType,
     TupleType(TupleTypeSignature)
 }
 
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub struct TypeSignature {
-    atomic_type: AtomTypeIdentifier,
+struct ListTypeData {
+    max_len: u32,
     dimension: u8
 }
 
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub struct TupleTypeSignature {
-    type_map: BTreeMap<String, TypeSignature>
+pub struct TypeSignature {
+    atomic_type: AtomTypeIdentifier,
+    list_dimensions: Option<ListTypeData>,
+    // NOTE: for the purposes of type-checks and cost computations, list size = dimension * max_length!
+    //       high dimensional lists are _expensive_ --- use lists of tuples!
 }
 
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub struct TupleData {
-    pub type_signature: TupleTypeSignature,
+    type_signature: TupleTypeSignature,
     data_map: BTreeMap<String, Value>
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct BuffData {
+    data: Vec<u8>,
+    length: u32
+}
+
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub enum Value {
     Void,
     Int(i128),
     Bool(bool),
-    Buffer(Box<[char]>),
+    Buffer(BuffData),
     List(Vec<Value>, TypeSignature),
+    Principal(u8, Vec<u8>), // a principal is a version byte + hash160 (20 bytes)
```

Comment:
You can enforce this in the type system by using `[u8; 20]` instead of `Vec<u8>`